### PR TITLE
Refresh correct partial during refresh on drop

### LINKED ISSUE #2592: Refresh on drop can create outliers in continuous aggregates
The drop chunk, which can be called manually or by a retention policy on a hypertable, calls the refresh on all continuous aggregates of the hypertable for each chunk to drop. The refresh recomputes buckets, which fully covers a chunk to drop, if an invalidation is recorded for it. In the case when chunks and buckets are not align, a retention policy can result in animalities in aggregated values in continuous aggregates if the invalidations happen in the oldest bucket, which contains the start of the oldest dropped chunk in a retention policy call.

The following pictures illustrate the case when chunks and buckets are not aligned and a retention policy was called three times. For each call the recalculation buckets on dropped data is demonstrated.

Drop on refresh with cagg

The aggregates values in continuous aggregate might contain anomalies, e.g., if the aggregate operator was `sum` or `count`:

Anomalities after drop on refresh with cagg

Invalidation in old data is not common. However, if it is enough common, so it is important to refresh on drop, then the described anomaly case will be also not so uncommon. So if the refresh on drop is kept, then this issue needs to be solved too and the solution is not just switching from expanding buckets window to shrinking buckets window, since refresh is done per chunk, while several chunks can be dropped together.

### PR DESCRIPTION
**Pull request by k-rus**
When drop_chunks is called on a hypertable manually or through
a retention policy, it initiates refresh in hypertable's continuous
aggregates. The refresh includes all buckets of each dropped chunk.
Some buckets are inside a dropped chunk only partially. In such case,
this commit fixes to refresh only the correct partials of the buckets.

The fix passes chunk id to refresh on drop of a chunk. The chunk id is
used to identify and update correct partials. This is a workaround
that the invalidation outside of the refreshed chunk doesn't update
the other partial, which is outside the refreshed chunk.

The commit also adds the test demonstrating correct refresh.

Fixes #2592

### CODE REVIEW

File: tsl/src/continuous_aggs/materialize.c [hunk 1]
```diff
@@ -200,18 +205,27 @@ internal_time_range_to_time_range(InternalTimeRange internal)
...
-							Name time_column_name, TimeRange invalidation_range)
+							Name time_column_name, TimeRange invalidation_range,
+							const int32 chunk_id)
 {
-	spi_delete_materializations(materialization_table, time_column_name, invalidation_range);
+	StringInfo chunk_condition = makeStringInfo();
+
+	if (chunk_id != INVALID_CHUNK_ID)
+		appendStringInfo(chunk_condition, "AND chunk_id = %d", chunk_id);
+	spi_delete_materializations(materialization_table,
```
Comment (erimatnor) [inline tsl/src/continuous_aggs/materialize.c:215]: I recommend a newline between if clause and the following statements

File: tsl/test/expected/continuous_aggs_drop_chunks.out [hunk 2]
```diff
@@ -139,3 +139,136 @@ SELECT drop_chunks('records', '2000-03-16'::timestamptz);
  _timescaledb_internal._hyper_1_15_chunk
 (15 rows)
 
+\set VERBOSITY terse
+DROP MATERIALIZED VIEW records_monthly;
+NOTICE:  drop cascades to 32 other objects
+DROP TABLE records;
+DROP TABLE clients;
+\set VERBOSITY default
+-- Test how caggs are afected by drop_chunk due to refresh on drop
```
Comment (erimatnor) [inline tsl/test/expected/continuous_aggs_drop_chunks.out:148]:
```suggestion
-- Test how caggs are affected by drop_chunk due to refresh on drop
```

File: tsl/test/expected/continuous_aggs_drop_chunks.out [hunk 3]
```diff
@@ -139,3 +139,136 @@ SELECT drop_chunks('records', '2000-03-16'::timestamptz);
...
+     28 | 21.98 |     7
+     35 | 21.98 |     7
+     42 | 21.98 |     7
+     49 | 21.98 |     7
+     56 | 21.98 |     7
+     63 | 21.98 |     7
+(10 rows)
+
+-- This is the simplest case, when the updated bucket is inside chunk, so it is expected
+-- always reflected.
```
Comment (erimatnor) [inline tsl/test/expected/continuous_aggs_drop_chunks.out:197]: I don't understand this comment. Specifically, "...so it is expected always reflected"

File: tsl/test/expected/continuous_aggs_drop_chunks.out [hunk 4]
```diff
@@ -139,3 +139,136 @@ SELECT drop_chunks('records', '2000-03-16'::timestamptz);
...
+     28 | 21.98 |     7
+     35 | 21.98 |     7
+     42 | 21.98 |     7
+     49 | 21.98 |     7
+     56 | 21.98 |     7
+     63 | 21.98 |     7
+(10 rows)
+
+-- This case is update in a bucket, which crosses two chunks, which exist and both is going
+-- to be dropped.
```
Comment (erimatnor) [inline tsl/test/expected/continuous_aggs_drop_chunks.out:222]:
```suggestion
--  Test an UPDATE that affects a bucket that is split across two chunks. Both chunks are going to be dropped.
```

File: tsl/test/expected/continuous_aggs_drop_chunks.out [hunk 5]
```diff
@@ -139,3 +139,136 @@ SELECT drop_chunks('records', '2000-03-16'::timestamptz);
...
+UPDATE conditions SET value = 4.00 WHERE time_int = 31;
+-- This update is in the bucket, which crosses two chunks. The first chunk is going to be 
+-- dropped now and the update is outside it, and thus should not be reflected. The second 
+-- chunk contains the update and will be dropped on the next call, thus it should be reflected 
+-- in the continuous aggregate only after the last drop.
+UPDATE conditions SET value = 4.00 WHERE time_int = 41;
+-- Now the update in 31 will be present, but not the update in 41.
```
Comment (erimatnor) [inline tsl/test/expected/continuous_aggs_drop_chunks.out:229]:
```suggestion
-- After drop_chunks, which refreshes the cagg, the update in 31 will be visible, but not the update in 41.
```

File: tsl/test/expected/continuous_aggs_drop_chunks.out [hunk 6]
```diff
@@ -139,3 +139,136 @@ SELECT drop_chunks('records', '2000-03-16'::timestamptz);
...
+(10 rows)
+
+-- This case is update in a bucket, which crosses two chunks, which exist and both is going
+-- to be dropped.
+UPDATE conditions SET value = 4.00 WHERE time_int = 31;
+-- This update is in the bucket, which crosses two chunks. The first chunk is going to be 
+-- dropped now and the update is outside it, and thus should not be reflected. The second 
+-- chunk contains the update and will be dropped on the next call, thus it should be reflected 
+-- in the continuous aggregate only after the last drop.
+UPDATE conditions SET value = 4.00 WHERE time_int = 41;
```
Comment (erimatnor) [inline tsl/test/expected/continuous_aggs_drop_chunks.out:228]: Isn't a more interesting test case one where you update the _same_ bucket twice, but each update falls within a separate partial. Then when you drop the first chunk, the first partial of bucket X will be updated, but not the second partial of bucket X. For instance,

```
UPDATE conditions SET value = 4.00 WHERE time_int = 29;
UPDATE conditions SET value = 4.00 WHERE time_int = 32;
```

These two updates fall in the same bucket `[28,35)`, but are two different partials since the chunk border is at 30.

File: tsl/test/expected/continuous_aggs_drop_chunks.out [hunk 7]
```diff
@@ -139,3 +139,136 @@ SELECT drop_chunks('records', '2000-03-16'::timestamptz);
...
+     21 | 21.98 |     7
+     28 | 22.84 |     7
+     35 | 21.98 |     7
+     42 | 21.98 |     7
+     49 | 21.98 |     7
+     56 | 21.98 |     7
+     63 | 21.98 |     7
+(10 rows)
+
+-- Now the update the update in 41 will be also present.
```
Comment (erimatnor) [inline tsl/test/expected/continuous_aggs_drop_chunks.out:252]:
```suggestion
-- Now the update in 41 will be also be visible.
```

File: tsl/test/expected/continuous_aggs_drop_chunks.out (hunk 6 above)
Comment (k-rus) [reply tsl/test/expected/continuous_aggs_drop_chunks.out:228]: It is not more interested than the current tests. It doesn't allow to see if the fix recognises, which partials need to be updated, and this test was already working correctly before this fix.
I will add it.

Comment (erimatnor) [reply tsl/test/expected/continuous_aggs_drop_chunks.out:230]: Maybe I am missing something, but if a bucket can be split across two partials (one in each chunk), it seems reasonable to test updates in those two same-bucket partials to see that both cases are properly reflected in the bucket when dropping the corresponding chunks.

For instance, bucket 2 below is split across two chunks, thus consisting of two partials 2a and 2b. When dropping the first chunk we'd like to see that bucket 2 is updated to reflect changes in chunk 1 and when the second chunk is dropped we'd like to see that bucket 2 is updated to reflect changes in chunk 2.

```
  1    2     3
|----|-----|----| buckets
       a|b
|-------|-------| chunks
```

File: tsl/test/sql/continuous_aggs_drop_chunks.sql [hunk 8]
```diff
@@ -51,3 +51,67 @@ CALL refresh_continuous_aggregate('records_monthly', NULL, NULL);
...
+SELECT time_val, time_val % 4, 3.14 FROM generate_series(0,100,1) AS time_val;
+
+SELECT set_integer_now_func('conditions', 'test_int_now');
+CREATE MATERIALIZED VIEW conditions_7
+    WITH (timescaledb.continuous, timescaledb.materialized_only = TRUE)
+    AS
+        SELECT time_bucket(7, time_int) as bucket,
+            SUM(value), COUNT(value)
+        FROM conditions GROUP BY bucket WITH DATA;
+SELECT materialization_hypertable_schema||'.'||materialization_hypertable_name AS mat_hyper
```
Comment (erimatnor) [inline tsl/test/sql/continuous_aggs_drop_chunks.sql:80]: I don't think `mat_hyper` is ever used. Is this query used at all?

File: tsl/test/expected/continuous_aggs_drop_chunks.out [hunk 9]
```diff
@@ -139,3 +139,139 @@ SELECT drop_chunks('records', '2000-03-16'::timestamptz);
...
+    $$ SELECT 125 $$;
+CREATE TABLE conditions(time_int INT NOT NULL, device INT, value FLOAT);
+SELECT create_hypertable('conditions', 'time_int', chunk_time_interval => 10);
+    create_hypertable    
+-------------------------
+ (3,public,conditions,t)
+(1 row)
+
+INSERT INTO conditions
+SELECT time_val, time_val % 4, 3.14 FROM generate_series(0,100,1) AS time_val;
```
Comment (erimatnor) [inline tsl/test/expected/continuous_aggs_drop_chunks.out:160]: Would be nice to use the new chunks view to show the ranges of the chunks before dropping them so that one can easily compare the ranges to the buckets in the cagg.

File: tsl/test/expected/continuous_aggs_drop_chunks.out [hunk 10]
```diff
@@ -139,3 +139,139 @@ SELECT drop_chunks('records', '2000-03-16'::timestamptz);
...
+     42 | 21.98 |     7
+     49 | 21.98 |     7
+     56 | 21.98 |     7
+     63 | 21.98 |     7
+(10 rows)
+
+-- This is the simplest case, when the updated bucket is inside a chunk, so it is expected
+-- that the update is always reflected after the refresh on drop.
+UPDATE conditions SET value = 4.00 WHERE time_int = 2;
+-- This case updates values in the bucket, which affects two different partials in
```
Comment (erimatnor) [inline tsl/test/expected/continuous_aggs_drop_chunks.out:197]: The test case I was referring to previously was to drop these chunks separately, to see that the _same_ bucket is updated twice; first when dropping the first chunk, then updated again when dropping the second chunk.

File: tsl/test/expected/continuous_aggs_drop_chunks.out (hunk 6 above)
Comment (k-rus) [reply tsl/test/expected/continuous_aggs_drop_chunks.out:230]: The described case is already producing correct result before this fix because it was refreshing bucket 2 two times, so the test doesn't demonstrated the corner case described by the issue.
This PR fixes that 2b is refreshed during drop of chunk b, and 2a is refreshed during drop of chunk a.  The regression test was already testing the most interesting cases. I added the test as you requested.

File: tsl/test/sql/continuous_aggs_drop_chunks.sql (hunk 8 above)
Comment (k-rus) [reply tsl/test/sql/continuous_aggs_drop_chunks.sql:80]: Thank you! It is left over after I queries the materialized table in the view, but it was redundant after I improved tests to cover corner cases better. I will remove this.

File: tsl/test/expected/continuous_aggs_drop_chunks.out [hunk 11]
```diff
@@ -139,3 +139,139 @@ SELECT drop_chunks('records', '2000-03-16'::timestamptz);
...
+-----------------------------------------
+ _timescaledb_internal._hyper_3_65_chunk
+ _timescaledb_internal._hyper_3_66_chunk
+(2 rows)
+
+SELECT * FROM see_cagg;
+ bucket |  sum  | count 
+--------+-------+-------
+      0 | 22.84 |     7
+      7 |  23.7 |     7
```
Comment (erimatnor) [inline tsl/test/expected/continuous_aggs_drop_chunks.out:212]: If you drop the two first chunks separately, I would expect bucket 7 to first refresh to 22.84 and then when the second chunk is dropped it would be 23.7. Why not test that case as I suggested?

Comment (erimatnor) [reply tsl/test/expected/continuous_aggs_drop_chunks.out:209]: Ohh, I think this is tested below... sorry

File: tsl/test/expected/continuous_aggs_drop_chunks.out (hunk 10 above)
Comment (k-rus) [reply tsl/test/expected/continuous_aggs_drop_chunks.out:194]: No, your suggestion was to drop from the chunks, which are dropped together. See your comment with your example: https://github.com/timescale/timescaledb/pull/2613#discussion_r516608352

File: tsl/test/expected/continuous_aggs_drop_chunks.out (hunk 11 above)
Comment (k-rus) [reply tsl/test/expected/continuous_aggs_drop_chunks.out:209]: Your case is already tested and I commented two times that I am adding it. Please, check the code

**Review by erimatnor — APPROVED** [review]

File: tsl/test/expected/continuous_aggs_drop_chunks.out (hunk 10 above)
Comment (erimatnor) [reply tsl/test/expected/continuous_aggs_drop_chunks.out:194]: My suggestion was to change the test  so that the chunks are not dropped together. But I see that the requested case is now covered in a different way

Comment (k-rus) [reply tsl/test/expected/continuous_aggs_drop_chunks.out:194]: I was mislead by your example, which was in the bucket of the chunks dropped together. Yes, the other test was also added.

**Review by gayyappan — APPROVED** [review]
Comments requested.

File: tsl/src/continuous_aggs/materialize.c [hunk 12]
```diff
@@ -34,18 +34,20 @@ static Datum internal_to_time_value_or_infinite(int64 internal, Oid time_type,
...
-										TimeRange materialization_range);
+										TimeRange materialization_range,
+										const char *chunk_condition);
 
 void
 continuous_agg_update_materialization(SchemaAndName partial_view,
 									  SchemaAndName materialization_table, Name time_column_name,
 									  InternalTimeRange new_materialization_range,
-									  InternalTimeRange invalidation_range, int64 bucket_width)
+									  InternalTimeRange invalidation_range, int64 bucket_width,
```
[the remark this answers is not in the document]
Comment (k-rus) [reply tsl/src/continuous_aggs/materialize.c:49]: Added a comment
